proposals/2045-column-lineage-endpoint.md

```diff
@@ -0,0 +1,189 @@
+# Proposal: Column lineage endpoint proposal
+
+Author(s): @julienledem, @pawel-big-lebowski
+
+Created: 20022-08-18
+
+Dicussion: [column lineage endpoint issue #2045](https://github.com/MarquezProject/marquez/issues/2045)
+
+## Overview
+
+### Use cases
+ - Find the current upstream dependencies of a column. A column in a dataset is derived from columns in upstream datasets.
+ - See column-level lineage in the dataset level lineage when available.
+ - Retrieve point-in-time upstream lineage for a dataset or a column. What did the lineage look like yesterday compared to today?
+
+### Existing elements
+
+- OpenLineage defines a [column-level lineage facet]- (https://github.com/OpenLineage/OpenLineage/blob/main/spec/facets/ColumnLineageDatasetFacet.json). 
+- Marquez has a lineage endpoint `GET /api/v1/lineage` that returns the current lineage graph connected to a job or a dataset
+
+### Column lineage characteristics and general assumptions
+
+Column level lineage is a different lineage graph due to a different node granularity - kind of zoomed-in view of existing lineage. Instead of datasets and jobs being lineage graph nodes, each dataset field becomes a node. Additionally, there are edges between dataset fields, instead of datasets itself. Thus, enriching existing lineage with column lineage information would not be sufficient. That’s why we propose another API endpoint with column lineage graph.
+
+Upstream and downstream edges do have different characteristics. An output dataset is always produced by a single version of input dataset (one upstream), while a single input datset version can have multiple output dataset versions. Lineage graph can be then easily flooded by downstream subgraph which blurs the overall view. That's why we consider an upstream column lineage as a default one. Downstream lineage will be returned only when requested explicitly. 
+
+### New Elements
+
+We propose to add the following:
+
+ - Add column lineage to the lineage endpoint as a part of dataset.
```

Comment:
I'm curious whether this won't result in much slower query time. Maybe column level lineage should be explicitely specified somehow on UI, and not returned always every time lineage is requested.

Reply:
I considered this as an extra foreign key `JOIN` which should not make it **much slower**.
Column lineage is one of many parts of dataset response. It may be an overkill to let user define what should be contained in response (at least for this PR), while adding an option to return/not-return only for a column lineage would be inconsistent with other facets. 

If `lineage` endpoint seems too rich already, I would remove column lineage from it and include it in DatasetResource, so that those who want it could fetch it with an extra call.

Reply:
Summing up offline discussion with @pawel-big-lebowski - the column lineage from main endpoint is unlikely to be often used on itself in comparison with one on the separate endpoint. For example, UI would not display data received by this by default, so it would be unused, yet always received from the service.